File: day1-foundations/lab1-bare-metal-loop/python/fixture_server.py
"""Tiny local HTTP server so the http_get tool has something deterministic to
call. Real endpoint, real HTTP, no dependency on the classroom network.

Safe to call repeatedly: if the port is already serving, it reuses it.
"""
import functools, http.server, pathlib, socket, threading

PORT = 8137
ROOT = pathlib.Path(__file__).resolve().parent.parent / "fixtures"
_server = None


def _port_is_open(port: int) -> bool:
    with socket.socket() as probe:
        probe.settimeout(0.3)
        return probe.connect_ex(("127.0.0.1", port)) == 0


def serve_in_background(port: int = PORT):
    """Start the fixture server unless something is already on the port."""
    global _server
    if _server is not None or _port_is_open(port):
        return _server
    handler = functools.partial(http.server.SimpleHTTPRequestHandler, directory=str(ROOT))
    handler.log_message = lambda *a, **k: None          # quiet
    http.server.HTTPServer.allow_reuse_address = True
    _server = http.server.HTTPServer(("127.0.0.1", port), handler)
    threading.Thread(target=_server.serve_forever, daemon=True).start()
    return _server


if __name__ == "__main__":
    serve_in_background()
    print(f"serving {ROOT} on http://127.0.0.1:{PORT}")
    threading.Event().wait()

File: day1-foundations/lab1-bare-metal-loop/node/fixture-server.mjs
// Local fixture server so http_get has something deterministic to call.
// Safe to call repeatedly.
import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import net from "node:net";
import { fileURLToPath } from "node:url";

export const PORT = 8137;
const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "fixtures");
let server = null;

function portIsOpen(port) {
  return new Promise((resolve) => {
    const socket = net.connect({ port, host: "127.0.0.1" });
    socket.setTimeout(300);
    socket.on("connect", () => { socket.destroy(); resolve(true); });
    socket.on("error", () => resolve(false));
    socket.on("timeout", () => { socket.destroy(); resolve(false); });
  });
}

export async function serveInBackground(port = PORT) {
  if (server || (await portIsOpen(port))) return server;
  server = http.createServer((req, res) => {
    const file = path.join(ROOT, path.basename(new URL(req.url, "http://x").pathname));
    if (!fs.existsSync(file)) { res.statusCode = 404; return res.end("not found"); }
    res.setHeader("content-type", "application/json");
    res.end(fs.readFileSync(file));
  });
  await new Promise((resolve) => server.listen(port, "127.0.0.1", resolve));
  server.unref();
  return server;
}
